### PR TITLE
registration exception, user entity, user dto, registration impl

### PR DESCRIPTION
**Pull request by artemsa223**

### CODE REVIEW

File: src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    @Email
+    private String email;
+    @NotBlank
+    @Length(min = 8, max = 35)
+    private String password;
+    @Length(min = 8, max = 35)
+    private String repeatPassword;
+    private String firstName;
+    private String lastName;
+    private String shippingAddress;
```
Comment (MentorMA) [inline src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java]: Feel free for validation this fields

File: src/main/java/com/example/demo/repository/UserRepository.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.example.demo.repository;
+
+import com.example.demo.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
```
Comment (MentorMA) [inline src/main/java/com/example/demo/repository/UserRepository.java:8]: Do you need to customize this method, it should be implemented inside JPA, just use it

File: src/main/java/com/example/demo/validation/fieldmatch/FieldMatchValidator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        boolean valid = true;
+        try {
+            final Object firstObj = BeanUtils.getProperty(value, firstFieldName);
+            final Object secondObj = BeanUtils.getProperty(value, secondFieldName);
+
+            valid = firstObj == null && secondObj == null
+                    || firstObj != null && firstObj.equals(secondObj);
+        } catch (final Exception ignore) {
+            // ignore
+        }
```
Comment (MentorMA) [inline src/main/java/com/example/demo/validation/fieldmatch/FieldMatchValidator.java]: Looks not good, you can add some log or smth else

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/repository/UserRepository.java (hunk 2 above)
Comment (artemsa223) [reply src/main/java/com/example/demo/repository/UserRepository.java:8]: I didn`t find that  method inside JPA

File: src/main/java/com/example/demo/controller/AuthController.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            throws RegistrationException {
+        return userService.registerUser(requestDto);
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    public void deleteById(@PathVariable Long id) {
+        userService.deleteById(id);
+    }
+
```
Comment (okuzan) [inline src/main/java/com/example/demo/controller/AuthController.java]:
```suggestion
```

File: src/main/java/com/example/demo/controller/AuthController.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/auth")
+public class AuthController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
```
Comment (okuzan) [inline src/main/java/com/example/demo/controller/AuthController.java]:
```suggestion
    @PostMapping("/register")
```
imo it's more common and simpler to type

File: src/main/java/com/example/demo/controller/AuthController.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/auth")
```
Comment (okuzan) [inline src/main/java/com/example/demo/controller/AuthController.java]: extract /api prefix to `server.servlet.context-path`

File: src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    private String password;
+    @Length(min = 8, max = 35)
+    private String repeatPassword;
+    @NotBlank
+    private String firstName;
+    @NotBlank
+    private String lastName;
+    @NotBlank
+    private String shippingAddress;
+
```
Comment (okuzan) [inline src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+@FieldMatch(first = "password", second = "repeatPassword",
+        message = "The password fields must match")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Email
+    private String email;
+    @NotBlank
+    @Length(min = 8, max = 35)
+    private String password;
+    @Length(min = 8, max = 35)
```
Comment (okuzan) [inline src/main/java/com/example/demo/dto/user/UserRegistrationRequestDto.java]:
```suggestion
    @NotNull
    @Length(min = 8, max = 35)
    private String password;
    @NotNull
    @Length(min = 8, max = 35)
```
NotBlank is possible too, but we have a more specific validator Length already, IMO notnull communicates intent clearer here. And you didn't add any validation for repeatPassword. Remember, if we pass null as a field annotated with Length/Size it won't fail

File: src/main/java/com/example/demo/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -40,4 +41,21 @@ private String getErrorsMessage(ObjectError e) {
...
+    }
+
+    @ExceptionHandler({EntityNotFoundException.class})
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException
+                                                                               exception) {
+        return ResponseEntity
+                .status(HttpStatus.NOT_FOUND)
+                .body(exception.getMessage());
+    }
+
```
Comment (okuzan) [inline src/main/java/com/example/demo/exception/CustomGlobalExceptionHandler.java]:
```suggestion
```

File: src/main/java/com/example/demo/dto/user/UserResponseDto.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package com.example.demo.dto.user;
+
+import lombok.Data;
+
+@Data
+public class UserResponseDto {
```
Comment (okuzan) [inline src/main/java/com/example/demo/dto/user/UserResponseDto.java:6]: not a mistake but why not use record for your DTOs

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 11]
```diff
@@ -1,3 +1,5 @@
 databaseChangeLog:
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: don't forget that one empty line at EOF

File: src/main/java/com/example/demo/service/user/UserService.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.demo.dto.user.UserRegistrationRequestDto;
+import com.example.demo.dto.user.UserResponseDto;
+import com.example.demo.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto registerUser(UserRegistrationRequestDto requestDto)
+            throws RegistrationException;
+
+    void deleteById(Long id);
+
```
Comment (okuzan) [inline src/main/java/com/example/demo/service/user/UserService.java]:
```suggestion
```

File: src/main/java/com/example/demo/model/User.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+@Table(name = "users")
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true,name = "email")
+    private String email;
+    @Column(nullable = false,name = "first_name")
+    private String firstName;
+    @Column(nullable = false,name = "last_name")
```
Comment (okuzan) [inline src/main/java/com/example/demo/model/User.java]: always format your files with IDEA's formatter (Ctrl Alt L to trigger it)
there should be a space after each comma

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/dto/user/UserResponseDto.java (hunk 10 above)
Comment (artemsa223) [reply src/main/java/com/example/demo/dto/user/UserResponseDto.java:6]: just easier to work and understand for me

**Pull request merged**